config/env_spec_request.go

```diff
@@ -0,0 +1,309 @@
+// Create using NewEnvironmentSpecRequest()
+type EnvironmentSpecRequest struct {
+	*EnvironmentSpecExt
+	request            *authv3.CheckRequest
+	authMan            auth.Manager
+	jwtAuthentications map[string]*JWTAuthentication // JWTAuthentication.Name ->
+	jwtResults         map[string]*jwtResult         // JWTAuthentication.Name ->
+	verifier           jwt.Verifier
+	apiSpec            *APISpec
+	operation          *APIOperation
+	queryValues        url.Values
+}
+
+type jwtClaims map[string]interface{}
+
+type jwtResult struct {
+	claims jwtClaims
+	err    error
+}
+
+// GetJWTResult returns the claims and error if a JWTAuthentication of the passed name was
+// verified, nil if it was not verified or does not exist
+func (e *EnvironmentSpecRequest) GetJWTResult(name string) (map[string]interface{}, error) {
+	if e != nil {
+		if jwtResult := e.jwtResults[name]; jwtResult != nil {
+			return jwtResult.claims, jwtResult.err
+		}
+	}
+	return nil, nil
+}
+
+func (e *EnvironmentSpecRequest) GetAPISpec() *APISpec {
+	if e == nil {
+		return nil
+	}
+	if e.apiSpec == nil {
+		path := strings.Split(e.getRequestPath(), "/")
+		if result := e.apiPathTree.Find(path, 0); result != nil {
+			e.apiSpec = result.(*APISpec)
+		}
+	}
+	return e.apiSpec
+}
+
+// path without querystring
+func (e *EnvironmentSpecRequest) getRequestPath() string {
+	return strings.SplitN(e.request.Attributes.Request.Http.Path, "?", 2)[0]
+}
+
+// GetOperation uses HttpMatch to return an APIOperation
+func (e *EnvironmentSpecRequest) GetOperation() *APIOperation {
+	if e == nil {
+		return nil
+	}
+	if e.operation == nil {
+		api := e.GetAPISpec()
+		if api != nil {
+			subPath := strings.TrimPrefix(e.getRequestPath(), api.BasePath) // strip base path
+			pathSplits := strings.Split(subPath, "/")
+			// prepend method for search
+			pathSplits = append([]string{e.apiSpec.ID, e.request.Attributes.Request.Http.Method}, pathSplits...)
+
+			if result := e.opPathTree.Find(pathSplits, 0); result != nil {
+				e.operation = result.(*APIOperation)
+			}
+		}
+	}
+	return e.operation
+}
+
+// GetParamValue extracts a value from request using Match
+func (e *EnvironmentSpecRequest) GetParamValue(param APIOperationParameter) string {
+	if e == nil || param.Match == nil {
+		return ""
+	}
+	var value string
+	switch m := param.Match.(type) {
+	case Header:
+		value = e.request.Attributes.Request.Http.Headers[strings.ToLower(string(m))]
+		// Per Envoy: If multiple headers share the same key, they are merged per HTTP spec.
+		// So, we're just grabbing the first value (up to any comma).
+		if indx := strings.Index(value, ","); indx > 0 {
+			value = value[:indx]
+		}
+	case Query:
+		if e.queryValues == nil {
+			q := strings.SplitN(e.request.Attributes.Request.Http.Path, "?", 2)
+			if len(q) > 1 {
+				vals, _ := url.ParseQuery(q[1])
+				e.queryValues = vals
+			}
+			value = e.queryValues.Get(string(m))
```

Comment:
I haven't fully examined but just to make sure this is purposefully within the `if {}` block?

Reply:
Yup, it's just a lazy initialization of e.queryValues.